fix:  'open in new tab' option issue & multiple link control UX issues

hz-tyfoon opened the pull request:
[This is the updated version of the PR: #50401 that maintains the current state of the `link control` & add the fixs on it]
Fix issues: https://github.com/WordPress/gutenberg/issues/45741 & https://github.com/WordPress/gutenberg/issues/43144
Also addresses some UI/UX issues mentioned here:  [this comment](https://github.com/WordPress/gutenberg/issues/49091#issuecomment-1477057235), also [this comment with a video demonstration ](https://github.com/WordPress/gutenberg/issues/49091#issuecomment-1477084979) & [this comment](https://github.com/WordPress/gutenberg/issues/49091#issuecomment-1477097292) 

*** What I've Done 👇👇 ***
### Update: `4` & `5` in the list below has been removed/reverted from this PR, check https://github.com/WordPress/gutenberg/pull/50602#issuecomment-1550356168
** 1) took a new state 'newValue'/'setNewValue' and took 'value' it's initial default. and replaced the usage of 'value' with this 'newValue' state everywhere in the index.js to make this state the source of truth. we'll need this state for couple of reasons and one being 'updating' the value using 'setNewValue' & another is replace 'onChange' call with 'setNewValue' or use both of them according to our need (more on this in the next steps to make it clear why this new state is taken, see the next steps for more clarity)

** 2) instead of passing the 'onChange' as prop to 'LinkControlSettingsDrawer', passed 'setNewValue' as prop. Because, the 'onChange' call makes the link popup modal to close without saving the other data properly. In 'settings-drawer.js' the receieved 'setNewValue' passed to the 'Settings' component as 'onChange' prop, this way toggling the 'open in new tab' won't close the modal & also will save the data in 'newValue' state properly.

** 3) in 'handleSubmit' of 'index.js' removed the if check as looks like it's unnecessary. Because if user decide to submit without making any change to 'url' or 'title' there's no harm in still submitting it, also there's other change can happen like 'open in new tab settings' and user may just wanna submit that. Another thing done it 'handleSubmit' & that is called both 'setNewValue' and 'onChnage' to make sure the 'local state' (newValue) stays sync with the submitted data (otherwise there may be unwanted behaviour where 'link-control component' is used. i.e. 'button block')

** 4) imported the 'Settings' component inside 'index.js' and used it below the 'LinkPreview' component. used the neccessary css in 'style.scss' to properly style it. In this 'Settings' component's 'onChange' prop, passed a function & made sure the value gets called with both 'setNewValue' and 'onChnge'. This is done for the same reason of keeping the both state sync. This settings component showen there to keep the 'open in new tab option' visible when previewing the link (after initially any richtext is linked by selecting the text & then pasting the link ). This was done to keep it consistent with 'WordPress 6.2 core UI' and looks like many people find it better UX to keep the 'option' there. See these comments for more info: [this comment](https://github.com/WordPress/gutenberg/issues/49091#issuecomment-1477057235), also [this comment with a video demonstration ](https://github.com/WordPress/gutenberg/issues/49091#issuecomment-1477084979) & [this comment](https://github.com/WordPress/gutenberg/issues/49091#issuecomment-1477097292)

** 5) finally 2 extra works done related to keeping the settings panel visible when opening the modal as seems like it's a better UX to keep this option visible rather than keeping them hidden and then requiring 1 extra click to make them open. To achieve this set the initial default value to 'true' for 'settingsOpen' useState & removed the 'setSettingsOpen(false)' from 'stopEditing' function.

*** Thanks 😊😊 ***

<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

hz-tyfoon committed:
fix: multiple link control isues

*** What I've Done 👇👇 ***

** 1) took a new state 'newValue'/'setNewValue' and took 'value' it's initial default. and replaced the usage of 'value' with this 'newValue' state everywhere in the index.js to make this state the source of truth. we'll need this state for couple of reasons and one being 'updating' the value using 'setNewValue' & another is replace 'onChange' call with 'setNewValue' or use both of them according to our need (more on this in the next steps to make it clear why this new state is taken, see the next steps for more clarity)

** 2) instead of passing the 'onChange' as prop to 'LinkControlSettingsDrawer', passed 'setNewValue' as prop. Because, the 'onChange' call makes the link popup modal to close without saving the other data properly. In 'settings-drawer.js' the receieved 'setNewValue' passed to the 'Settings' component as 'onChange' prop, this way toggling the 'open in new tab' won't close the modal & also will save the data in 'newValue' state properly.

** 3) in 'handleSubmit' of 'index.js' removed the if check as looks like it's unnecessary. Because if user decide to submit without making any change to 'url' or 'title' there's no harm in still submitting it, also there's other change can happen like 'open in new tab settings' and user may just wanna submit that. Another thing done it 'handleSubmit' & that is called both 'setNewValue' and 'onChnage' to make sure the 'local state' (newValue) stays sync with the submitted data (otherwise there may be unwanted behaviour where 'link-control component' is used. i.e. 'button block')

** 4) imported the 'Settings' component inside 'index.js' and used it below the 'LinkPreview' component. used the neccessary css in 'style.scss' to properly style it. In this 'Settings' component's 'onChange' prop, passed a function & made sure the value gets called with both 'setNewValue' and 'onChnge'. This is done for the same reason of keeping the both state sync. This settings component showen there to keep the 'open in new tab option' visible when previewing the link (after initially any richtext is linked by selecting the text & then pasting the link ). This was done to keep it consistent with 'WordPress 6.2 core UI' and looks like many people find it better UX to keep the 'option' there. See these comments for more info: 1) https://github.com/WordPress/gutenberg/issues/49091\#issuecomment-1477057235 2) https://github.com/WordPress/gutenberg/issues/49091\#issuecomment-1477084979 (see the video in this comment for more clarity) 3) https://github.com/WordPress/gutenberg/issues/49091\#issuecomment-1477097292

** 5) finally 2 extra works done related to keeping the settings panel visible when opening the modal as seems like it's a better UX to keep this option visible rather than keeping them hidden and then requiring 1 extra click to make them open. To achieve this set the initial default value to 'true' for 'settingsOpen' useState & removed the 'setSettingsOpen(false)' from 'stopEditing' function.

*** Thanks 😊😊 ***

hz-tyfoon commented:
Here's a video demo of the Fixs:

https://github.com/WordPress/gutenberg/assets/59374562/3ff0aec6-8dce-4fae-98d7-6bfa316746f5

cc: @getdave @annezazu @supernovia @jordesign 
Feel free to share your thoughts/reviews/feedbacks. Thanks 😊 🙏

draganescu commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -203,29 +205,24 @@ function LinkControl( {
+			title: internalTextInputValue,
+		};
+		setNewValue( valueToSubmit );
+		onChange( valueToSubmit );
```
The original code didn't trigger onChange if nothing changed. Is this still true?

hz-tyfoon replied:
Thanks for the review 😊 🙏 @draganescu 
When submitted, Here I triggered onChange even the the 'title' & 'url' wasn't changed. I described the following in my PR's description that's I believe related to your question:


> ** 3) in 'handleSubmit' of 'index.js' removed the if check as looks like it's unnecessary. Because if user decide to submit without making any change to 'url' or 'title' there's no harm in still submitting it, also there's other change can happen like 'open in new tab settings' and user may just wanna submit that.

hz-tyfoon replied:
Just an update: I haven't found any issue caused by this so far. I'm assuming U haven't too. 
But if U see any `issue` or `potential issue/problem` with this changes please feel free to share. 

I'm not resolving this conversation at this moment. But looks so far it can be resolved if no issue found from your end.
Thanks again @draganescu . 😊

draganescu commented:
I tested this and stumbled upon this apparently wrong behavior:


https://github.com/WordPress/gutenberg/assets/107534/c55acdb3-709e-4fe3-b8ed-ae463dd78936

Clicking "Cancel" should close LinkUI with no changed saved in no state.

hz-tyfoon commented:
thanks again @draganescu for catching it 🙏 . Think I found where it's coming from. Just, little occupied now but, will push a fix soon I get a chance.

hz-tyfoon committed:
fix: clicking 'cancel' button still shows the updated change

 - fixed: when making some change after opening the 'link control modal' & without applying/submitting it, when the 'cancel' button is clicked it show's that the changes already applied even though it's cancelled (although the updated changes aren't really applied but, still it's a misleading UI so fixed it)

hz-tyfoon commented:
> I tested this and stumbled upon this apparently wrong behavior:
> 
> Clicking "Cancel" should close LinkUI with no changed saved in no state.

@draganescu 
Fixed the issue👆you shared in this commit: 2b35bad . Can U check & confirm please? 🙏 
Thanks a lot for testing the PR & finding it 😊

getdave commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -203,29 +205,24 @@ function LinkControl( {
 			...updatedValue,
 			title: internalTextInputValue || updatedValue?.title,
 		} );
-		stopEditing();
```
My instinct is that this is straying into changing the behaviour of the control too much.

The current behaviour is selecting a suggestion submits it. I don't think this PR should look to change that.

getdave commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -331,15 +322,27 @@ function LinkControl( {
-					hasUnlinkControl={ shownUnlinkControl }
-					onRemove={ onRemove }
-				/>
+			{ newValue && ! isEditingLink && ! isCreatingPage && (
```
This whole section is for previewing a _submitted_ link. We should never need the stored internal `newValue` at this point. Rather the preview should preview the _controlled_ link `value` that is provided to the component.

hz-tyfoon replied:
I think `newValue` is better here as it'll be same as the 'value' if nothing changed or the 'updated value' if anything changed. 
But, U're right in the sense that it's not needed here at this point and we can disscuss about it separately from this PR. So reverted it" 05924eb29b

getdave commented on packages/block-editor/src/components/link-control/settings-drawer.js:
```diff
@@ -28,7 +28,7 @@ function LinkSettingsDrawer( {
 	value,
 	settings,
-	onChange,
+	setNewValue = () => {},
```
We could reuse the existing `onChange` prop here but pass in a different callback. No need to add the new `setNewValue` prop.

getdave commented on packages/block-editor/src/components/link-control/style.scss:
```diff
@@ -462,13 +462,21 @@ $preview-image-height: 140px;
 	padding: $grid-unit-20;
+}
+
+.block-editor-link-control__tools_in_preview {
+	border-top: 1px solid #ddd;
+}
+
+.block-editor-link-control__tools {
+	flex-wrap: wrap;
+	justify-content: space-between;
```
Thank you for looking at visual design here. However, it's really important we try to keep PRs small and focused. This makes it easier to roll back changes that cause regressions and also makes it easier to review PRs.

I think we should remove everything from this PR that doesn't fix [the bug ](https://github.com/WordPress/gutenberg/issues/45741). 

Feel free to spin out other PRs with the other changes.

Apologies if this seems annoying or pedantic but it's proven to be an important principle of working in a large open source codebase. Thank you 🙇‍♂️

getdave reviewed (commented):
Thank you for continuing to iterate on this.

I think we need to reduce the number of changes in this PR to keep it focused on doing a single thing. It's great to have your help in addressing so many issues but it makes the PR harder to review and also makes it harder to identify and rollback PRs that cause regressions. In general we advocate for smaller PRs that do one thing. Let's remove `4` and `5` as a minimum - feel free to spin out other PRs if you are able.

I've left some initial comments. Don't feel you have to address those immediately. I'm going to dive in more today and look at this in more detail. Appreciate your patience and continued work here 🙇

hz-tyfoon committed:
reverted 'fix/item: 4' (see the description for details) as looks like it's better to handle it separately from 'PR 50602'

*** What was the 'fix/item: 4' that's reverted in this commit? here's what I wrote about it in my description 👇👇 ***

"imported the 'Settings' component inside 'index.js' and used it below the 'LinkPreview' component. used the neccessary css in 'style.scss' to properly style it. In this 'Settings' component's 'onChange' prop, passed a function & made sure the value gets called with both 'setNewValue' and 'onChnge'. This is done for the same reason of keeping the both state sync. This settings component showen there to keep the 'open in new tab option' visible when previewing the link (after initially any richtext is linked by selecting the text & then pasting the link ). This was done to keep it consistent with 'WordPress 6.2 core UI' and looks like many people find it better UX to keep the 'option' there. See these comments for more info: this comment, also this comment with a video demonstration & this comment"

hz-tyfoon committed:
reverted 'fix/item: 5' (see the description for details) as looks like it's better to handle it separately from 'PR 50602'

*** What was the 'fix/item: 5' that's removed in this commit? here's what I wrote about it in my description 👇👇 ***

"finally 2 extra works done related to keeping the settings panel visible when opening the modal as seems like it's a better UX to keep this option visible rather than keeping them hidden and then requiring 1 extra click to make them open. To achieve this set the initial default value to 'true' for 'settingsOpen' useState & removed the 'setSettingsOpen(false)' from 'stopEditing' function."

hz-tyfoon committed:
revert / (slight refactor): kept the prop name 'onChange' rather than 'setNewValue' but passed the 'setNewValue function' as suggested by dave

I inititally renamed both the prop 'name and value' to 'setNewValue' so that it prevents any confusion being created (like, is the recieved 'onChange' prop in 'settings-drawe.js' is the same one which was recieved in the 'index.js'?) But, looks like that's not something to worry about at this moment. Looks like it's better to have that disscussion separately from this PR

hz-tyfoon commented:
Thanks again @getdave for the explanation & the review. 😊 
Really sorry 😞 about this 1 PR addressing multiple things. After your explanation looks like it's better to keep concerns separate from this PR. 

> I think we need to reduce the number of changes in this PR to keep it focused on doing a single thing. It's great to have your help in addressing so many issues but it makes the PR harder to review and also makes it harder to identify and rollback PRs that cause regressions. In general we advocate for smaller PRs that do one thing. Let's remove `4` and `5` as a minimum - feel free to spin out other PRs if you are able.

I removed `4` & `5` in [c5eb7dc](https://github.com/WordPress/gutenberg/pull/50602/commits/c5eb7dc4a452b387c0840edbe88102797f2a60be) & [e3ce2c6](https://github.com/WordPress/gutenberg/pull/50602/commits/e3ce2c61ca25c23a93d145f431216642c78ba01a) respectively.

I'll create separate `issue/PR` for both of them later.

Thanks. 😊 
Checking your other reviews as well.

hz-tyfoon committed:
revert / (slight refactor): reverted back to old behaviour (selecting suggested links closes the modal) as per review

Why?

To keep concerns separate. Think It'd be better if this is disscussed separately & this PR shouldn't look to change this behaviour. more by dave here: https://github.com/WordPress/gutenberg/pull/50602\#discussion_r1194818569

hz-tyfoon committed:
revert: used 'value' rather than 'newValue' as per review

why?

We can disscuss which one is better in this case separately from this PR. see this https://github.com/WordPress/gutenberg/pull/50602\#discussion_r1194821669

hz-tyfoon commented:
I addressed all of your reviews so far, can U check 🙏 Thanks.

getdave commented:
Thank you. On my list to review...

getdave commented:
Let's take a moment to consider the proposed UX/UI refresh of Link Control in https://github.com/WordPress/gutenberg/issues/50891 before we decide how to proceed here.

richtabor commented:
@getdave is this still relevant, given the current state of link control?

getdave commented:
I believe all the Issues described here are resolved by our work in upstream repo. Safe close out.
